server/api/orphaned-record/orphaned-record.controller.ts

```diff
@@ -0,0 +1,242 @@
+        .getOne();
+
+      if (rosterHistory) {
+        // Backdate the timestamp to the earliest orphaned record time.
+        rosterHistory.timestamp = orphanedRecords.map(({ timestamp }) => timestamp).sort()[0];
+        rosterHistory.save();
+      } else {
+        // Add a new roster entry
+        addRosterEntry(req.appOrg!, req.appUserRole!.role, req.body);
+      }
+
+      for (const orphanedRecord of orphanedRecords) {
+        // Remove the orphan record entry
+        const newOrphanedRecord = await orphanedRecord.remove();
+
+        // Request a reingestion of a single document
+        const queryInput = makeDocumentIdQuery(orphanedRecord.documentId);
+        const reingestResult = await reingest(queryInput);
+
+        // Delete any outstanding actions
+        await OrphanedRecordAction
+          .createQueryBuilder()
+          .delete()
+          .where(`id=:id`, { id: req.params.id })
+          .orWhere('expires_on < now()')
+          .execute();
+
+        // Track the individual ingestion and postgres save() results
+        resultRecords.push({
+          ...reingestResult,
+          orphanedRecord: newOrphanedRecord,
+        });
+
+        // Keep track of our aggregate totals
+        lambdaInvocationCount += reingestResult.lambdaInvocationCount;
+        recordsIngested += reingestResult.recordsIngested;
+      }
+    });
+
+    res.json({
+      items: resultRecords,
+      lambdaInvocationCount,
+      recordsIngested,
+    });
+  }
+
+  async addOrphanedRecordAction(req: ApiRequest<OrphanedRecordActionParam, OrphanedRecordActionData>, res: Response) {
+    if (!req.params.id) {
+      throw new BadRequestError(`Param 'id' is required.`);
+    }
+    if (req.params.id.split(';').length !== 4) {
+      throw new BadRequestError(`Param 'id' is malformed. ${req.params.id}`);
+    }
+    if (!req.body.action) {
+      throw new BadRequestError(`Expected 'action' in payload.`);
+    }
+    if (!(req.body.timeToLiveMs > 0)) {
+      throw new BadRequestError(`Expected 'timeToLiveMs' in payload.`);
+    }
+
+    const existingAction = await OrphanedRecordAction.findOne({
+      where: {
+        user_edipi: req.appUser.edipi,
+        id: req.params.id,
+      },
+    });
+    if (existingAction) {
+      await existingAction.remove();
+    }
+
+    const orphanedRecords = await OrphanedRecord
+      .createQueryBuilder('orphan')
+      .where('orphan.joinKey=:joinKey', { joinKey: req.params.id })
+      .getMany();
+
+    if (orphanedRecords.length === 0) {
+      throw new BadRequestError(`Unable to locate orphaned record with id: ${req.params.id}`);
+    }
+
+    // 1231231231;17035813720;unit;1
+
+    const now = Date.now();
+    const orphanedRecordAction = new OrphanedRecordAction();
+    orphanedRecordAction.id = req.params.id;
+    orphanedRecordAction.type = req.body.action;
+    orphanedRecordAction.user = req.appUser;
+    orphanedRecordAction.createdOn = new Date(now);
```

Comment:
I don't think you need to set `createdOn` since it's already a generated column.